### PR TITLE
Nano: Add automatically rescale learning rate

### PR DESCRIPTION
**Pull request by y199387**
## Description
Add automatically learning rate configuration during multi-instance training. The strategy will scale the lr of optimizers' param_groups and base_lrs of lr_schedulers by the number of processes. Linear warmup will be applied on the optimizers if no other lr_schedulers is set.
- linear learning rate scaling: The lr of each optimizers' param_groups and base_lrs of each lr_scheduler will be scaled by the number of processes, for example:
```python
class Mymodel(pl.LightningModule):
...
def configure_optimizers(self):
    optimizer = SGD(
        self.parameters(),
        lr=0.05
    )
    lr_scheduler = LinearLR(
        optimizer,
        total_iters=2
    )
trainer = Trainer(num_processes=2, max_epochs=10)
model = Mymodel()
trainer.fit(model, train_dataloaders=train_loader)
```
The lr of optimizer's param_groups[0] will be scaled to 0.1 from 0.05, the same for lr_scheduler's base_lrs.
- Linear warmup strategy: The `start_lr` is the original learning rate, the `end_lr` is the scaled learning rate and the `warmup_epochs` is set to `max_epochs // 10` if `max_epochs >= 10` otherwise `max_steps //10` by default if the value isn't given in `auto_lr`. For the code above without lr_scheduler, start_lr=0.05, end_lr=0.1, warmup_epochs=1. The change of learning rate is shown in the figure below:
image

### 1. Why the change?
Improve performance in multi-instance training.
related issue
https://github.com/intel-analytics/BigDL/issues/4855

### 2. User API changes
The flag `auto_lr` defaults to True. In multi-instance training, the above operations are preformed by default.
```python
from bigdl.nano.pytorch import Trainer
trainer = Trainer(num_processes=2,
                           distributed_backend='subprocess,
                           auto_lr=True,    # optionally
                           max_epochs=5)
trainer.fit(model, train_dataloaders = train_loader)
```
Meanwhile `auto_lr` can also accept an dict to control `warmup_epochs`, or it is set to max_epochs // 10 by default.
```python
from bigdl.nano.pytorch import Trainer
trainer = Trainer(num_processes=2,
                           distributed_backend='subprocess',
                           auto_lr={'warmup_epochs': 3},
                           max_epochs=10)
trainer.fit(model, train_dataloaders=train_loader)
```

### 3. How to test?
- [x] Unit test
- [x] Jenkins

### CODE REVIEW

**Comment by jason-dai** [comment]
Shall we also add warmup @yangw1234 ?

**Comment by yangw1234** [comment]
> Shall we also add warmup @yangw1234 ?

Yes, I think we should add warmup. However, adding warmup may require users to specify warmup steps or epochs.  We may redesign the API.

**Comment by jason-dai** [comment]
> > Shall we also add warmup @yangw1234 ?
> 
> Yes, I think we should add warmup. However, adding warmup may require users to specify warmup steps or epochs. We may redesign the API.

We should always provide reasonable default values; BTW, is `scale_lr` default to true?

**Comment by y199387** [comment]
> 

`scale_lr` is default to false for now.

**Comment by y199387** [comment]
> > > Shall we also add warmup @yangw1234 ?
> > 
> > 
> > Yes, I think we should add warmup. However, adding warmup may require users to specify warmup steps or epochs. We may redesign the API.
> 
> We should always provide reasonable default values; BTW, is `scale_lr` default to true?
`scale_lr` is default to false for now.

**Comment by y199387** [comment]
> > > Shall we also add warmup @yangw1234 ?
> > 
> > 
> > Yes, I think we should add warmup. However, adding warmup may require users to specify warmup steps or epochs. We may redesign the API.
> 
> We should always provide reasonable default values; BTW, is `scale_lr` default to true?

scale_lr is default to false for now.

**Comment by jason-dai** [comment]
> > > > Shall we also add warmup @yangw1234 ?
> > > 
> > > 
> > > Yes, I think we should add warmup. However, adding warmup may require users to specify warmup steps or epochs. We may redesign the API.
> > 
> > 
> > We should always provide reasonable default values; BTW, is `scale_lr` default to true?
> 
> scale_lr is default to false for now.

I think the goal is to make it default to true, maybe after warmup is added? @yangw1234

File: python/nano/test/pytorch/tests/test_plugin.py [hunk 1]
```diff
@@ -112,7 +112,8 @@ def test_trainer_subprocess_correctness(self):
             loss=torch.nn.MSELoss(),
             metrics=[torchmetrics.MeanSquaredError()]
         )
-        trainer = Trainer(num_processes=2, distributed_backend="subprocess", max_epochs=2)
+        trainer = Trainer(num_processes=2, distributed_backend="subprocess", max_epochs=2,
+                          auto_lr=False)
```
Comment (y199387) [inline python/nano/test/pytorch/tests/test_plugin.py:116]: Keep the learning rate constant to test correctness.

**Comment by y199387** [comment]
Jenkins
http://10.112.231.51:18888/job/BigDL-NanoTests-PR-Validation/329/

File: python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py [hunk 2]
```diff
@@ -221,6 +227,68 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
+                for opt_idx, opt in enumerate(self.optimizers):
+                    if type(self.auto_lr) is bool:
+                        self.auto_lr = trainer.max_epochs // 5 + 1
+                    if not TORCH_VERSION_LESS_1_10:
+                        from torch.optim.lr_scheduler import LinearLR
+                        scheduler = LinearLR(optimizer=opt,
+                                             # set initial lr as user lr
+                                             start_factor=1.0 / self.world_size,
+                                             end_factor=1.0,
+                                             total_iters=self.auto_lr)
```
Comment (yangw1234) [inline python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py:261]: Does `total_iters` mean the number of epochs?

Comment (y199387) [reply python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py:261]: Yes

File: python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py [hunk 3]
```diff
@@ -221,6 +227,68 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
+                for param_group in optimizer.param_groups:
+                    param_group["lr"] *= self.world_size
+
+            lr_scheduler_configs = self.lr_scheduler_configs
+            for config in lr_scheduler_configs:
+                scheduler = config.scheduler
+                scheduler.base_lrs = [lr * self.world_size for lr in scheduler.base_lrs]
+
+            if self.lr_scheduler_configs:
+                warnings.warn(f"Nano warmup currently only support none optimizers, "
```
Comment (yangw1234) [inline python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py:248]: I am a little confused about this warning. What do you mean by "none optimizers" and how do you "skip warmup"?

Comment (y199387) [reply python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py:248]: "none optimizers" should be "no lr_scheduler", I will fix it.

File: python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py [hunk 4]
```diff
@@ -221,6 +227,68 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
+                              f"but got {len(self.lr_scheduler_configs)}. Skip warmup")
+            else:
+                lr_schedulers = []
+                for opt_idx, opt in enumerate(self.optimizers):
+                    if type(self.auto_lr) is bool:
+                        self.auto_lr = trainer.max_epochs // 5 + 1
```
Comment (yangw1234) [inline python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py:254]: Is there any reference or reasoning for setting this strategy?

File: python/nano/test/pytorch/tests/test_scale_lr.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+import torch.nn.functional as F
+
+num_classes = 10
+batch_size = 32
+dataset_size = 256
+num_workers = 0
+data_dir = os.path.join(os.path.dirname(__file__), "../data")
+
+
+class Resnet_2_0(pl.LightningModule):
```
Comment (yangw1234) [inline python/nano/test/pytorch/tests/test_scale_lr.py:42]: What's the difference between `Resnet_2_0` and `Resnet_2_2`? Is it possible to use more meaningful names?

File: python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py [hunk 6]
```diff
@@ -221,6 +227,68 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
+                    }
+                    lr_schedulers.append(lr_scheduler)
+                lr_scheduler_configs = (
+                    _configure_schedulers_automatic_opt(lr_schedulers, None)
+                    if self.lightning_module.automatic_optimization
+                    else _configure_schedulers_manual_opt(lr_schedulers)
+                )
+                _set_scheduler_opt_idx(self.optimizers, lr_scheduler_configs)
+                _validate_scheduler_api(lr_scheduler_configs, self.model)
+                self.lr_scheduler_configs = lr_scheduler_configs
```
Comment (yangw1234) [inline python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py:290]: Are these lines adapted from some original code in the pytorch-lightning code base?

Comment (yangw1234) [reply python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py:290]: Maybe add some reference?

Comment (y199387) [reply python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py:290]: Yes, they are used to make sure the lr_scheduler are
established correctly

File: python/nano/test/pytorch/tests/test_scale_lr.py (hunk 5 above)
Comment (y199387) [reply python/nano/test/pytorch/tests/test_scale_lr.py:42]: Resnet_2_0 used to test the optimizers' lr scaling and warmup has two optimizers and no lr_scheduler.
And Resnet_2_2 used to test lr_schedulers' base_lrs scaling

File: python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py (hunk 6 above)
Comment (y199387) [reply python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py:290]: https://github.com/Lightning-AI/lightning/blob/1.6.4/pytorch_lightning/core/optimizer.py#L189

File: python/nano/test/pytorch/tests/test_scale_lr.py (hunk 5 above)
Comment (y199387) [reply python/nano/test/pytorch/tests/test_scale_lr.py:42]: I'm not sure, maybe `Resnet_opt` and `Resnet_opt_sch`?

Comment (yangw1234) [reply python/nano/test/pytorch/tests/test_scale_lr.py:42]: `ResNetWith2Optimzers` and `ResNetWithScheduler`?

File: python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py (hunk 4 above)
Comment (y199387) [reply python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py:254]: Actually, I didn't find a specific suggestion on how to set warm_epochs. The warmup_epochs is 4000 and max_epochs is 40000 in some transformer implemetation. In facebook's paper, warmup_epoch is set to 5 and max_epochs is 80.

File: python/nano/test/pytorch/tests/test_scale_lr.py (hunk 5 above)
Comment (y199387) [reply python/nano/test/pytorch/tests/test_scale_lr.py:42]: ok, these are better.

**Comment by y199387** [comment]
jenkins
http://10.112.231.51:18889/job/BigDL-Chronos-PR-Validation/639/
http://10.112.231.51:18888/job/ZOO-PR-nano-pytorch-ipex-test/304/

**Comment by yangw1234** [comment]
LGTM

**Comment by yangw1234** [comment]
you may also need to change the description in https://bigdl.readthedocs.io/en/latest/doc/Nano/QuickStart/pytorch_train_quickstart.html
